File: src/services/searchService.ts
// src/services/searchService.ts

import api from './api';

export interface SearchResultDetails {
  specialization?: string;
  experience?: string;
  reviews?: number;
  education?: string;
  languages?: string[];
  hospital?: string;
  visitingHours?: string;
  departments?: string[];
  parking?: string;
  genericName?: string;
  dosage?: string;
  form?: string;
  quantity?: string;
  sideEffects?: string[];
}

export interface SearchResult {
  id: string;
  title: string;
  description: string;
  type: 'doctor' | 'hospital' | 'medicine' | 'pharmacy';
  link: string;
  image?: string;
  rating?: number;
  location?: string;
  distance?: string;
  price?: string;
  availability?: string;
  details?: SearchResultDetails;
  fee?: string;
  reviews?: number;
  beds?: number;
  emergency?: string;
  manufacturer?: string;
  prescription?: boolean;
  pharmacies?: Array<{ name: string; price: string; distance: string; inStock: boolean }>;
  substitutes?: string[];
}

class SearchService {
  async searchDoctors(query: string, userLocation?: { lat: number; lng: number }): Promise<SearchResult[]> {
    const response = await api.get<any>('/doctors', { search: query, limit: 10 });
    if (response.success && response.data) {
      const items = Array.isArray(response.data) ? response.data : response.data?.data || [];
      return items.map((doc: any) => ({
        id: String(doc.id),
        title: doc.full_name || doc.name || `Dr. ${doc.specialization}`,
        specialization: doc.specialization,
        description: `⭐ ${doc.rating || 0} (${doc.review_count || 0} reviews) • ${doc.experience_years || 0} years exp • ${doc.specialization || ''}`,
        type: 'doctor' as const,
        link: `/doctor/${doc.id}`,
        rating: doc.rating || 0,
        location: doc.hospital_affiliation || '',
        distance: '',
        availability: doc.status === 'online' ? 'Available Now' : 'Offline',
        fee: doc.consultation_fee ? `$${doc.consultation_fee}` : '',
        details: {
          specialization: doc.specialization,
          experience: `${doc.experience_years || 0} years`,
          reviews: doc.review_count,
          education: doc.qualifications ? JSON.parse(doc.qualifications)[0]?.institution : '',
          languages: doc.languages ? JSON.parse(doc.languages) : [],
          hospital: doc.hospital_affiliation || '',
        },
      }));
    }
    return [];
  }

  async searchHospitals(query: string, userLocation?: { lat: number; lng: number }): Promise<SearchResult[]> {
    const response = await api.get<any>('/hospitals', { search: query, limit: 10 });
    if (response.success && response.data) {
      const items = Array.isArray(response.data) ? response.data : response.data?.data || response.data?.hospitals || [];
      return items.map((h: any) => ({
        id: h.id,
        title: h.name,
        description: `🏥 ${h.rating || 0} (${h.review_count || 0} reviews) • ${h.total_beds || 0} beds • ${h.type || ''}`,
        type: 'hospital' as const,
        link: `/hospital/${h.id}`,
        rating: h.rating || 0,
        location: `${h.city || ''}, ${h.state || ''}`,
        distance: '',
        reviews: h.review_count,
        beds: h.total_beds,
        emergency: h.emergency_service || '',
        details: {
          departments: h.services ? JSON.parse(h.services) : [],
          visitingHours: '',
          parking: '',
        },
      }));
    }
    return [];
  }

  async searchMedicines(query: string): Promise<SearchResult[]> {
    const response = await api.get<any>('/pharmacy/medicines', { search: query, limit: 10 });
    if (response.success && response.data) {
      const items = Array.isArray(response.data) ? response.data : response.data?.data || [];
      return items.map((m: any) => ({
        id: String(m.id),
        title: m.medicine_name || m.name || '',
        description: `${m.category || ''} • ${m.dosage || ''} • ${m.form || ''}`,
        type: 'medicine' as const,
        link: `/pharmacy/medicine/${m.id}`,
        rating: m.rating || 0,
        price: m.price ? `$${m.price}` : '',
        details: {
          genericName: m.generic_name || '',
          dosage: m.dosage || '',
          form: m.form || '',
          quantity: m.pack_size || '',
          sideEffects: m.side_effects ? JSON.parse(m.side_effects) : [],
        },
        manufacturer: m.manufacturer || '',
        prescription: m.requires_prescription || false,
      }));
    }
    return [];
  }

  async globalSearch(query: string, userLocation?: { lat: number; lng: number }): Promise<SearchResult[]> {
    if (!query.trim()) return [];

    const [doctors, hospitals, medicines] = await Promise.all([
      this.searchDoctors(query, userLocation),
      this.searchHospitals(query, userLocation),
      this.searchMedicines(query),
    ]);

    return [...doctors, ...hospitals, ...medicines]
      .sort((a, b) => (b.rating || 0) - (a.rating || 0))
      .slice(0, 15);
  }
}

export const searchService = new SearchService();

File: services/search_service.py
from sqlalchemy.orm import Session
from sqlalchemy import or_
from typing import Optional
from datetime import datetime

from ..models.doctor import DoctorProfile
from ..models.hospital import Hospital
from ..models.pharmacy import Pharmacy, PharmacyInventory
from ..models.user import User
from ..models.search import SearchLog
from ..schemas.search import SearchResultItem, SearchResponse, SearchLogResponse
from ..utils.helpers import build_pagination_meta


class SearchService:
    def __init__(self, db: Session):
        self.db = db

    def global_search(
        self, query: str, user_id: Optional[str] = None,
        search_type: Optional[str] = None,
        page: int = 1, size: int = 20,
    ) -> SearchResponse:
        results = []
        term = f"%{query}%"

        if not search_type or search_type == "doctor":
            doctor_results = (
                self.db.query(DoctorProfile, User)
                .join(User, User.id == DoctorProfile.user_id)
                .filter(
                    DoctorProfile.is_active == True,
                    User.is_active == True,
                    or_(
                        User.full_name.ilike(term),
                        DoctorProfile.specialization.ilike(term),
                        DoctorProfile.hospital_affiliation.ilike(term),
                    ),
                )
                .offset(0).limit(size)
                .all()
            )
            for doc, user in doctor_results:
                results.append(SearchResultItem(
                    id=doc.id,
                    title=user.full_name or "Doctor",
                    type="doctor",
                    description=f"{doc.specialization} • {doc.experience_years or 0} yrs exp",
                    rating=float(doc.rating or 0),
                    location=doc.hospital_affiliation,
                    price=float(doc.consultation_fee or 0),
                    availability=doc.status,
                    image=doc.profile_image,
                    extra={"specialization": doc.specialization, "is_verified": doc.is_verified},
                ))

        if not search_type or search_type == "hospital":
            hospital_results = (
                self.db.query(Hospital)
                .filter(
                    Hospital.is_active == True,
                    or_(
                        Hospital.name.ilike(term),
                        Hospital.city.ilike(term),
                        Hospital.state.ilike(term),
                    ),
                )
                .offset(0).limit(size)
                .all()
            )
            for h in hospital_results:
                results.append(SearchResultItem(
                    id=h.id,
                    title=h.name,
                    type="hospital",
                    description=f"{h.type or 'Hospital'} • {h.city}, {h.state}",
                    rating=float(h.rating or 0),
                    location=f"{h.city}, {h.state}",
                    availability=h.emergency_service,
                    extra={"total_beds": h.total_beds, "available_beds": h.available_beds, "is_verified": h.is_verified},
                ))

        if not search_type or search_type == "pharmacy":
            pharmacy_results = (
                self.db.query(Pharmacy)
                .filter(
                    Pharmacy.is_active == True,
                    or_(
                        Pharmacy.name.ilike(term),
                        Pharmacy.city.ilike(term),
                    ),
                )
                .offset(0).limit(size)
                .all()
            )
            for p in pharmacy_results:
                results.append(SearchResultItem(
                    id=p.id,
                    title=p.name,
                    type="pharmacy",
                    description=f"Pharmacy • {p.city}",
                    rating=float(p.rating or 0),
                    location=f"{p.city}, {p.state}",
                    extra={"is_24x7": p.is_24x7, "delivery_available": p.delivery_available},
                ))

        if not search_type or search_type == "medicine":
            medicine_results = (
                self.db.query(PharmacyInventory)
                .filter(
                    PharmacyInventory.is_available == True,
                    or_(
                        PharmacyInventory.medicine_name.ilike(term),
                        PharmacyInventory.generic_name.ilike(term),
                        PharmacyInventory.brand_name.ilike(term),
                    ),
                )
                .offset(0).limit(size)
                .all()
            )
            for m in medicine_results:
                results.append(SearchResultItem(
                    id=m.id,
                    title=m.medicine_name,
                    type="medicine",
                    description=f"{m.generic_name or ''} • {m.category or ''}",
                    price=float(m.discounted_price or m.price or 0),
                    image=m.image_url,
                    extra={"form": m.form, "requires_prescription": m.requires_prescription, "pharmacy_id": m.pharmacy_id},
                ))

        # Sort by rating desc
        results.sort(key=lambda x: x.rating or 0, reverse=True)

        # Paginate
        total = len(results)
        pages = (total + size - 1) // size if size > 0 else 0
        paginated = results[(page - 1) * size: page * size]

        # Log search
        if user_id:
            log = SearchLog(
                user_id=user_id,
                query=query,
                type=search_type,
                results_count=total,
            )
            self.db.add(log)
            self.db.commit()

        return SearchResponse(
            query=query,
            results=paginated,
            total=total,
            page=page,
            size=size,
            pages=pages,
        )

    def get_search_history(self, user_id: str, page: int = 1, size: int = 20) -> dict:
        query = self.db.query(SearchLog).filter(
            SearchLog.user_id == user_id,
        ).order_by(SearchLog.created_at.desc())

        total = query.count()
        logs = query.offset((page - 1) * size).limit(size).all()

        return {
            "items": [SearchLogResponse.model_validate(l) for l in logs],
            **build_pagination_meta(total, page, size),
        }
